cmd/payload/extract.go

```diff
@@ -0,0 +1,111 @@
+// SPDX-FileCopyrightText: 2025 k0s authors
+// SPDX-License-Identifier: Apache-2.0
+
+package payload
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/k0sproject/k0s/cmd/internal"
+	internallog "github.com/k0sproject/k0s/internal/pkg/log"
+	"github.com/k0sproject/k0s/pkg/assets"
+	"github.com/k0sproject/k0s/pkg/constant"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+func NewExtractCmd() *cobra.Command {
+	var dataDir string
+	debugFlags := (&internal.DebugFlags{}).LongRunning()
+
+	cmd := &cobra.Command{
+		Use:   "extract",
+		Short: "Extract embedded binaries",
+		Args:  cobra.NoArgs,
+		Long: `Extract all embedded binaries to the bin directory.
+
+This command extracts all binaries embedded in the k0s executable to the
+bin directory. By default, binaries are extracted to /var/lib/k0s/bin.
+
+When a custom data-dir is specified, binaries are extracted to <data-dir>/bin.
+
+This is useful for pre-staging binaries before starting k0s, especially in
+airgapped environments or when the k0s binary is stored on a read-only
+filesystem.
+
+Examples:
+  # Extract to default location (/var/lib/k0s/bin)
+  k0s payload extract
+
+  # Extract to custom location
+  k0s payload extract --data-dir /custom/path
+`,
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			// Initialize logging
+			internallog.InitLogging()
+			debugFlags.Run(cmd, args)
+			return nil
+		},
```

Comment:
The logging initialization happens in `main.go`. This is how other subcommands are doing it:

```suggestion
		PersistentPreRun: debugFlags.Run,
```